### PR TITLE
interface: add settings screen

### PR DESCRIPTION
**Pull request by liam-fitzgerald**
Adds a settings screen which allows for customisable backgrounds, the hiding of avatars and nicknames and the order of tiles on the homescreen. Includes affordances for configuring S3, and logging out. Persists state to localStorage. 

[demo](http://ewr1.vultrobjects.com/urbit/Screen%20Recording%202020-07-31%20at%205.03.35%20pm.mov)

Screen Shot 2020-07-31 at 5 10 03 pm

Screen Shot 2020-07-31 at 5 08 27 pm

### CODE REVIEW

File: pkg/interface/src/reducers/local.ts [hunk 1]
```diff
@@ -1,18 +1,36 @@
...
-import { LocalUpdate } from '../types/local-update';
+import { LocalUpdate, BackgroundConfig } from '../types/local-update';
 
-type LocalState = Pick<StoreState, 'sidebarShown' | 'selectedGroups' | 'dark' | 'baseHash'>;
+type LocalState = Pick<StoreState, 'sidebarShown' | 'selectedGroups' | 'baseHash' | 'hideAvatars' | 'hideNicknames' | 'background' | 'dark'>;
 
 export default class LocalReducer<S extends LocalState> {
+    rehydrate(state: S) {
+      try {
+        const json = JSON.parse(localStorage.getItem('localReducer') || '');
```
Comment (tacryt-socryp) [inline pkg/interface/src/reducers/local.ts:11]: Clever and good approach to storing visual settings.

File: pkg/interface/src/themes/light.ts [hunk 2]
```diff
@@ -162,7 +162,7 @@ const theme = {
   ],
   // z-index
   zIndices: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
-  breakpoints: ["768px", "1024px", "1440px", "2200px"],
+  breakpoints: ["34.375em", "46.875em", "60em"]
```
Comment (tylershuster) [inline pkg/interface/src/themes/light.ts:165]: yes

**Comment by matildepark** [comment]
Dojo needs its background made transparent.

image

**Comment by matildepark** [comment]
Groups, on the other hand, has too much transparency.

image

**Comment by matildepark** [comment]
It would be dope to let the user set "light / dark / auto" theme settings (I find the dark mode looks dope with the background, for eg.) but I think we need to wrap refactoring into indigo-react first because it's not simple to override `prefers-color-scheme`...

**Comment by matildepark** [comment]
image

This is a weird way to do this ... @urcades do we have a "drag and drop" design anywhere?

**Comment by matildepark** [comment]
image
This is a good place for letting the user upload a background via s3 — see the avatar upload in Groups for an example ...

**Comment by matildepark** [comment]
Finally — the clock uses some `transparent` colours that look bizarre with backgrounds.

image

Now that we `matchMedia` you'll want to pass the dark boolean to the clock to set `white` vs. the `gray0` backgrounds, instead of just falling back to transparents...

**Comment by tylershuster** [comment]
Re @matildepark comments on transparency perhaps checking for `@supports(background-filter: blur())` and then making a frosted pane effect and falling back to a semi-transparent one

**Comment by liam-fitzgerald** [comment]
cc: @urcades thoughts?
Screen Shot 2020-08-04 at 12 18 43 pm

**Comment by urcades** [comment]
> image
> 
> This is a weird way to do this ... @urcades do we have a "drag and drop" design anywhere?

We do not, but I'm happy to make one using existing parts

**Comment by urcades** [comment]
> cc: @urcades thoughts?
> Screen Shot 2020-08-04 at 12 18 43 pm

I can see this looking pleasant with a refreshed sidebar look and feel, but we'd need to tweak the translucency/blurring precisely to get visibility/accessibility/readability correct. Even with this example, the green text buttons can get a little washed out and hard to read – I'd only imagine this gets accentuated further with more vibrant background images.

We can revisit this as we assemble a new sidebar structure — transparency should be something we allow users to set parametrically at some point, possibly — Might be good to @g-a-v-i-n here, as I'd imagine we'd want to think of a wholesale way to begin introducing transparency/translucency into indigo, so we aren't doing this piecemeal.

**Comment by urcades** [comment]
So tl;dr, let's keep that sidebar background boring and white100 for the time being — we'll think about how to go about this in a systematic manner.

**Comment by tylershuster** [comment]
I apologize in advance if this comes off as a bit rambling, but allowing the user to customize settings is like feeding the proverbial mouse.

The cookie is allowing the user to change their background. Obviously, and intentionally, this changes the whole feel of the interface. There's no guaranteeing one look going with that background. So keeping everything white is the safe way to go.

Of course, they didn't choose the background because they want it to be hidden, so you give them some translucency settings. But again as Ed points out, the same setting won't work for everyone.

It seems to me that inherent to the capacity to change the interface is a set of *semantically variable* classes, as opposed to the current *utility* classes. E.g. `.bg-translucent` where the particular properties of that class can be set via interface. This too is a cookie because how much control do you give? Do you go full MySpace or do you just give them a slider to control `n` properties of `x` classes?

tl;dr I agree with Ed and thank you for coming to my ted talk

File: pkg/interface/src/themes/light.ts (hunk 2 above)
Comment (g-a-v-i-n) [reply pkg/interface/src/themes/light.ts:165]: hang on, why did we go with ems here?

**Comment by g-a-v-i-n** [comment]
willing to think about transparent backgrounds a bit more but agree with ed.

**Comment by tacryt-socryp** [comment]
No unnecessary bells and whistles should be a policy from our team as a whole.

File: pkg/interface/src/themes/light.ts (hunk 2 above)
Comment (liam-fitzgerald) [reply pkg/interface/src/themes/light.ts:165]: to match our current breakpoints in indigo-static

Comment (matildepark) [reply pkg/interface/src/themes/light.ts:165]: Actually @liam-fitzgerald brought up out of band that there is no way to do anything responsively with a minimum of 768 also, this seems like a valid concern.

**Comment by liam-fitzgerald** [comment]
Screen Shot 2020-08-05 at 1 46 43 pm

[drag and drop demo](http://ewr1.vultrobjects.com/urbit/dnd-demo.mov)

File: pkg/interface/src/themes/light.ts (hunk 2 above)
Comment (g-a-v-i-n) [reply pkg/interface/src/themes/light.ts:165]: Ok, I need to check on this because it seems counter to the [styled-system documentation](https://styled-system.com/responsive-styles), unless I'm misunderstanding these docs. I'd also prefer we use px for breakpoints. You can just multiply ems by 16 to get their px values.

Comment (tylershuster) [reply pkg/interface/src/themes/light.ts:165]: Not exactly. Really it should be rems but at this point there won't be any divergence. I always use rems because that way the interface scales in proportion to whatever the user sets their font size at. But I don't know what kind of guidelines this project follows. I've actually transitioned to setting font-size in millimeters at root and everything else to rem which bases it off that. That grounds it in physical reality.

File: pkg/interface/src/apps/profile/components/lib/DisplayForm.tsx [hunk 3]
```diff
@@ -0,0 +1,230 @@
+import React, { useCallback } from "react";
+
```
Comment (tacryt-socryp) [inline pkg/interface/src/apps/profile/components/lib/DisplayForm.tsx:2]: Should be broken into individual files.

File: pkg/interface/src/apps/profile/components/lib/S3Form.tsx [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+  Text,
+  Menu,
+  MenuButton,
+  MenuList,
+  MenuItem,
+} from "@tlon/indigo-react";
+
+import { Formik, Form } from "formik";
+import * as Yup from "yup";
+import GlobalApi from "../../../../api/global";
```
Comment (tacryt-socryp) [inline pkg/interface/src/apps/profile/components/lib/S3Form.tsx:19]: Should be broken into individual components

**Comment by matildepark** [comment]
Time to resolve some conflicts! @liam-fitzgerald 

@urcades Do we have a gear icon or something for a settings button? Post-omnibox there's no way to access settings.

**Comment by urcades** [comment]
One exists for export here:
https://www.figma.com/file/H1RAHV4KscSTnvrIiL0z8C/Indigo?node-id=312%3A147

This has come across in a meeting with Liam, but I'm unsure if it was
socialized more broadly: The mechanism of "System Settings" access should
effectively be via the user's profile — the idea is to conjoin the notion
of one's root identity, and one's ship settings into one location. "You"
are effectively "the ship", and so it seems to make sense to have settings
reside as a section within one's "profile".

Additionally, settings should totally be searchable in the omnibox.

—
~fabled-faster
https://urbit.org/

**Comment by matildepark** [comment]
Yes it should! Just remember to stick that in the index in lib/omnibox.js @liam-fitzgerald.

Yes, his initial go at this had the settings linked by clicking your sigil, which is no longer in the header bar anywhere...

—
~haddef-sigwen
https://urbit.org

**Comment by matildepark** [comment]
Like, it could be *exclusively* accessed by leaping to "Profile", but seems unfriendly?

**Comment by urcades** [comment]
Yeah, I totally see this not being legible at first bat. A part of me
thinks that editing the terminology from "Profile" to "Ship" would fix
this, but we'd be reifying one of our weird neologisms into
top-of-interface status.

—
~fabled-faster
https://urbit.org/

**Comment by liam-fitzgerald** [comment]
@urcades that link is dead, do you have another?

**Comment by tacryt-socryp** [comment]
I like the idea of keeping the word Profile but having a settings cog button to the right of it that's always present. Seems fine.

**Comment by matildepark** [comment]
An icon in the text? I vote for sigil icon, “Profile and Settings” copy.

**Comment by tacryt-socryp** [comment]
I meant a separate cog button that directs to settings. "Profile and settings" seems too heavy to me...

**Comment by matildepark** [comment]
Oh you’re talking about the header. I mean as an omnibox entry. The profile and settings are one thing is the issue.

**Comment by tacryt-socryp** [comment]
Is this ready?

**Comment by matildepark** [comment]
Awaiting a UX change and then I’ll do a basic test. I asked @urcades to dump a definitive mock here so we can wrap it.

**Comment by urcades** [comment]
@matildepark  + @liam-fitzgerald Apologies for the wait, needed to cross reference w/ gav + jimmy:

Screen Shot 2020-08-12 at 9 32 39 PM

The above are the two ways we can imagine "profile + settings" being embedded in a single "ship" button. We've gone over this quite a bit but we generally share the belief that most people will expect some configuration interface to exist where they see their ship name.

In addition, it'd be good to surface this same interface here, in the Leap menu:

Screen Shot 2020-08-12 at 11 02 49 PM

I realize the default Leap menu doesn't exist just yet as we shipped MVP, but we should consider the above menubar listing/capability to be surfaced in the default menu as well.

**Comment by liam-fitzgerald** [comment]
Screen Shot 2020-08-17 at 5 42 13 pm
 

ready to merge!

Screen Shot 2020-08-17 at 5 42 25 pm

**Comment by matildepark** [comment]
Can you match [this](https://github.com/urbit/urbit/pull/3334#issuecomment-675075797)? Also conflict. Oof.

**Comment by matildepark** [comment]
I believe post-Chat refactor this was dropped as a prop. 
image

**Comment by matildepark** [comment]
hiding avatars and nicknames also appears to not be included in Links, Publish, full stop. All apps should include the boolean!

**Comment by matildepark** [comment]
This also needs to pull root identity out of Groups and put its 'contact card' editing into settings.

**Comment by urcades** [comment]
Yes — I'm not sure if it has immediate bearings in the way we're
structuring settings, but we should really be conceptually extracting the
idea of the "root profile" completely away from groups — it's no longer a
structure of groups, it is a standalone representation of the ship a user
is operating, it's external cloaking/presentation, etc.

To the various points raised about profile development and externalization
Galen has been bringing up recently, we should really consider the notion
of "one's profile" to be a very core structure that is externalized in
other applications, is the "prototype" object of representation so to speak.


—
~fabled-faster
https://urbit.org/

**Comment by matildepark** [comment]
As of #3337 merge, this is now unblocked.

**Comment by liam-fitzgerald** [comment]
ready to go! @matildepark

**Comment by matildepark** [comment]
You resolved the conflict, but did you address these:

> This is amazing. Except one thing blocking the merge ...
> 
> Screen Shot 2020-08-18 at 1 18 33 PM
> 
> Screen Shot 2020-08-18 at 1 18 41 PM

> I believe post-Chat refactor this was dropped as a prop.
> image

> hiding avatars and nicknames also appears to not be included in Links, Publish, full stop. All apps should include the boolean!

> This also needs to pull root identity out of Groups and put its 'contact card' editing into settings.
> 
> So remove:
> 
> image
> 
> And put it in this sidebar as "Your Identity" above "ship settings":
> 
> image
> 
> Same right-hand panel.

**Comment by liam-fitzgerald** [comment]
whoops, on it!

**Comment by matildepark** [comment]
Missing from profile-overlay

image

**Comment by matildepark** [comment]
People with nicknames set still have Inter applied even if nicknames are disabled.

image

**Comment by matildepark** [comment]
And Your Identity should come first.

image

**Comment by matildepark** [comment]
That seems to be it. Everything else looks slick and these are tiny changes.

**Comment by liam-fitzgerald** [comment]
done!

File: pkg/arvo/app/landscape/index.html [hunk 5]
```diff
@@ -24,6 +24,5 @@
     <script src="/~landscape/js/channel.js"></script>
     <script src="/~landscape/js/session.js"></script>
     <script src="/~landscape/js/bundle/index.f58fbbc4b037bb976a2a.js"></script>
-    <script src="https://sdk.amazonaws.com/js/aws-sdk-2.1.12.min.js"></script>
```
Comment (matildepark) [inline pkg/arvo/app/landscape/index.html:27]: I have no idea how or why this got dropped, but I restored it in next-userspace.
